### ISSUE TITLE
3.1 - Conditionally validating not working as expected

### ISSUE THREAD

**Issue by patrickconroy** [cakephp/cakephp#7441]
Hello. Say I have two fields in a modelless form, `input_parent` and `input_child`. If I want to validate and say that `input_child` should only be checked to not be blank if `input_parent` is the exact value `test`, it doesn't work and the `on` callable I send has no effect on anything.

```
# In a form..
$validator
    ->add('input_child', 'notBlank', [
        'rule' => 'notBlank',
        'on' => function ($context) {
            if (isset($context['data']['input_parent']) && $context['data']['input_parent'] === 'test') {
                 return true;
            }
            return false;
        }
    ]);

# Then..
$data = [
    'input_parent' => 'this is not a test',
    'input_child' => '',
];
// Returns 'This field cannot be left empty.' for input_child (this is bad)

$data = [
    'input_parent' => 'test',
    'input_child' => '',
];
// Returns 'This field cannot be left empty.' for input_child (this is fine)

$data = [
    'input_parent' => 'test',
    'input_child' => 'foo',
];
// No errors (fine)
```

If my `on` function always returns false, it still runs and my validation errors include `input_child`, saying it can't be empty. I'm pretty sure this is a bug, because if I set `input_child` to some value, the validation runs as expected.

This is all the same if I use the shorter-hand
```
$validator->allowEmpty('input_child', function ($context) {
    return false;
});
// Always checks emptiness, but it should never.
```

TLDR; - An empty child is valid *if* parent isn't `test`, but the validator never lets me have an empty child.

**Comment by patrickconroy** [comment]
Whoops sorry.. the shorthand works, I was inverting truthiness on the return.

However, should the `Validator::add` method work per what I have? Or were the `notEmpty` `allowEmpty` callables added because of the `add` method not being able to handle the logic?

**Comment by markstory** [comment]
Validation rules with `on` conditions should not be invoked if the on condition returns `false`. I'll see if I can reproduce the same issue.

**Comment by markstory** [comment]
I wasn't able to reproduce the issue you're having. For me callable 'on' conditions worked correctly.

```php
$validator->add('url', 'valid', [
    'rule' => 'url',
    'on' => function ($context) {
        return $context['data']['url'] !== 'test';
    }
]);
```

The following console code does what I was expecting:

```
>>> $t = \Cake\ORM\TableRegistry::get('Bookmarks');
=> App\Model\Table\BookmarksTable {#229
     +"registryAlias": "Bookmarks",
     +"table": "bookmarks",
     +"alias": "Bookmarks",
     +"entityClass": "App\Model\Entity\Bookmark",
     +"associations": [
       "users",
       "tags",
     ],
     +"behaviors": [
       "Timestamp",
     ],
     +"defaultConnection": "default",
     +"connectionName": "default",
   }
>>> $b = $t->newEntity(['url' => 'test']);
=> App\Model\Entity\Bookmark {#245
     +"url": "test",
     +"[new]": true,
     +"[accessible]": [
       "user_id" => true,
       "title" => true,
       "description" => true,
       "url" => true,
       "user" => true,
       "tags" => true,
       "end_date" => true,
     ],
     +"[dirty]": [
       "url" => true,
     ],
     +"[original]": [],
     +"[virtual]": [],
     +"[errors]": [],
     +"[repository]": "Bookmarks",
   }
>>> $b->url;
=> "test"

// While the following fails validation as expected.
>>> $e = $t->newEntity(['url' => 'derp']);
=> App\Model\Entity\Bookmark {#233
     +"[new]": true,
     +"[accessible]": [
       "user_id" => true,
       "title" => true,
       "description" => true,
       "url" => true,
       "user" => true,
       "tags" => true,
       "end_date" => true,
     ],
     +"[dirty]": [],
     +"[original]": [],
     +"[virtual]": [],
     +"[errors]": [
       "url" => [
         "valid" => "The provided value is invalid",
       ],
     ],
     +"[repository]": "Bookmarks",
   }
```
